File: packages/froxel-math/src/rect.ts

import { Module } from "asyl";
import Vec2 from './vec2';

let m: Module = null;

/**
 * Rectangle.
 */
export default class Rect
{
	/**
	 * Address of underlying Rect.
	 */
	addr: number;

	/**
	 * Underlying vector data.
	 */
	data: Float32Array;

	/**
	 * Binds the specified Asyl module to the class.
	 */
	static bind (module: Module) : void {
		m = module;
	}

	/**
	 * Allocates a new rectangle of zero size.
	 */
	static alloc() : Rect;
	/**
	 * Allocates a new rectangle with the specified size.
	 */
	static alloc (width: number, height: number, topLeft?: boolean) : Rect;
	/**
	 * Allocates a new rectangle with the specified coordinates.
	 */
	static alloc (x1: number, y1: number, x2: number, y2: number) : Rect;

	static alloc (x1: number=null, y1: number=null, x2: number|boolean=false, y2: number=null) : Rect
	{
		if (x1 === null)
			return new Rect(m.rect_alloc4f(0, 0, 0, 0));

		if (y2 === null)
			return new Rect(m.rect_alloc2f(x1, y1, x2));

		return new Rect(m.rect_alloc4f(x1, y1, x2, y2));
	}

	/**
	 * Materializes a rectangle at the specified memory location.
	 */
	static materialize(addr: number) : Rect
	{
		return new Rect(m.rect_materialize(addr));
	}

	/**
	 * Constructs the rectangle object.
	 */
	private constructor (addr: number) {
		this.addr = addr;
		this.data = m.mapFloat32Array(addr, 6);
	}

	/**
	 * Destroys the instance.
	 */
	dtor() : void {
		m.rect_dtor(this.addr);
	}

	/**
	 * Returns a clone of the rectangle.
	 */
	clone() : Rect {
		return new Rect(m.rect_clone(this.addr));
	}

	/**
	 * Sets all coordinates of the rectangle to zero.
	 */
	zero() : Rect {
		m.rect_zero(this.addr);
		return this;
	}

	/**
	 * Sets all coordinates of the rectangle to `null` for subsequent use with `extend`.
	 */
	reset() : Rect {
		m.rect_reset(this.addr);
		return this;
	}

	/**
	 * Extends the rectangle to contain the specified vector coordinates.
	 */
	extend (v: Vec2) : Rect;
	/**
	 * Extends the rectangle to contain the specified point.
	 */
	extend (x: number, y: number) : Rect;

	extend (x: Vec2|number, y: number=null) : Rect
	{
		if (y === null)
			m.rect_extend1v(this.addr, (x as Vec2).addr);
		else
			m.rect_extend2f(this.addr, x, y);

		return this;
	}

	/**
	 * Translates the rectangle by the vector coordinates.
	 */
	translate (v: Vec2) : Rect;
	/**
	 * Translates the rectangle by the given deltas.
	 */
	translate (dx: number, dy: number) : Rect;

	translate (dx: Vec2|number, dy: number=null) : Rect
	{
		if (dy === null)
			m.rect_translate1v(this.addr, (dx as Vec2).addr);
		else
			m.rect_translate2f(this.addr, dx, dy);

		return this;
	}

	/**
	 * Moves the center of the rectangle to the specified position.	 
	 * @param normalized - When `true` the parameters `x` and `y` are treated as normalized ranging from 0 to 1 (inclusive).
	 * !centerAt (x: number, y: number, normalized?: false) : Rect;
	 */
	center (x: number, y: number, normalized: boolean=false) : Rect
	{
		m.rect_center(this.addr, x, y, normalized);
		return this;
	}

	/**
	 * Copies the coordinates from the specified rectangle.
	 */
	set (r: Rect) : Rect;
	/**
	 * Sets the coordinates of the rectangle.
	 */
	set (x1: number, y1: number, x2: number, y2: number) : Rect;

	set (x1: Rect|number, y1: number=null, x2?: number, y2?: number) : Rect
	{
		if (y1 === null)
			m.rect_set1r(this.addr, (x1 as Rect).addr);
		else
			m.rect_set4f(this.addr, x1, y1, x2, y2);

		return this;
	}

	/**
	 * Returns `true` if the given rectangle coordinates are equal.
	 */
	equals (r: Rect) : boolean;
	/**
	 * 	Returns `true` if the coordinates are equal.
	 */
	equals (x1: number, y1: number, x2: number, y2: number) : boolean;
	
	equals (x1: Rect|number, y1: number=null, x2?: number, y2?: number) : boolean
	{
		if (y1 === null)
			return m.rect_equals1r(this.addr, (x1 as Rect).addr);

		return m.rect_equals4f(this.addr, x1, y1, x2, y2);
	}

	/**
	 * Returns `true` if the rectangle contains the given one.
	 */
	contains (r: Rect) : boolean;
	/**
	 * Returns `true` if the rectangle contains the given one describes by the (x1, y1) to (x2, y2) coordinates.
	 */
	contains (x1: number, y1: number, x2: number, y2: number) : boolean;
	/**
	 * Returns `true` if the rectangle contains the vector given coordinates.
	 */
	contains (v: Vec2, epsilon?: number) : boolean;
	/**
	 * Returns `true` if the rectangle contains the given coordinates.
	 */
	contains (x: number, y: number, epsilon?: number) : boolean;

	contains (x1: Rect|Vec2|number, y1: number=0, x2: number=0, y2: number=null) : boolean
	{
		if (x1 instanceof Rect)
			return m.rect_contains1r(this.addr, x1.addr);

		if (x1 instanceof Vec2)
			return m.rect_contains1v(this.addr, x1.addr, x2);

		if (y2 === null)
			return m.rect_contains2f(this.addr, x1, y1, x2);

		return m.rect_contains4f(this.addr, x1, y1, x2, y2);
	}

	/**
	 * Sets the coordinates of the rectangle to the union of it and the given one.
	 */
	union (r: Rect) : Rect;
	/**
	 * Sets the coordinates of the rectangle to the union of it and the given one.
	 */
	union (x1: number, y1: number, x2: number, y2: number) : Rect;
	
	union (x1: Rect|number, y1: number=null, x2?: number, y2?: number) : Rect
	{
		if (y1 === null)
			m.rect_union1r(this.addr, (x1 as Rect).addr);
		else
			m.rect_union4f(this.addr, x1, y1, x2, y2);

		return this;
	}

	/**
	 * Returns `true` if the rectangles intersect.
	 */
	intersects (r: Rect) : Rect;
	/**
	 * Returns `true` if the rectangle and the given coordinates intersect.
	 */
	intersects (x1: number, y1: number, x2: number, y2: number) : Rect;

	intersects (x1: Rect|number, y1: number=null, x2?: number, y2?: number) : Rect
	{
		if (y1 === null)
			m.rect_intersects1r(this.addr, (x1 as Rect).addr);
		else
			m.rect_intersects4f(this.addr, x1, y1, x2, y2);

		return this;
	}

	/**
	 * Sets the coordinates of the rectangle to the intersection of it and the given one.
	 */
	intersection (r: Rect) : boolean;
	/**
	 * Sets the coordinates of the rectangle to the intersection of it and the given one.
	 */
	intersection (x1: number, y1: number, x2: number, y2: number) : boolean;

	intersection (x1: Rect|number, y1: number=null, x2?: number, y2?: number) : boolean
	{
		if (y1 === null)
			return m.rect_intersection1r(this.addr, (x1 as Rect).addr);

		return m.rect_intersection4f(this.addr, x1, y1, x2, y2);
	}

	/**
	 * Resizes the rectangle to the given size using its center or top-left corner as reference.
	 * @param topLeft - When `true` reference will be top-left corner, set to `false` to use the center.
	 * @param normalized - When `true` the `width` and `height` will be treated as normalized values ranging from 0 to 1 (inclusive).
	 */
	resize (width: number, height: number, topLeft: boolean=false, normalized: boolean=false) : Rect
	{
		m.rect_resize (this.addr, width, height, topLeft, normalized);
		return this;
	}

	/**
	 * Resizes the rectangle using the specified deltas, relative to its center or top-left corner.
	 * @param topLeft - When `true` reference will be top-left corner, set to `false` to use the center.
	 */
	resizeBy (dWidth: number, dHeight: number, topLeft: boolean=false) : Rect
	{
		m.rect_resizeBy (this.addr, dWidth, dHeight, topLeft);
		return this;
	}

	/**
	 * Returns the value of the X1 coordinate.
	 */
	x1() : number {
		return this.data[0];
	}

	/**
	 * Returns the value of the Y1 coordinate.
	 */
	y1() : number {
		return this.data[1];
	}


	/**
	 * Returns the value of the X2 coordinate.
	 */
	x2() : number {
		return this.data[2];
	}

	/**
	 * Returns the value of the Y2 coordinate.
	 */
	y2() : number {
		return this.data[3];
	}

	/**
	 * Returns the value of the center X coordinate.
	 */
	cx() : number {
		return this.data[4];
	}

	/**
	 * Returns the value of the center Y coordinate.
	 */
	cy() : number {
		return this.data[5];
	}

	/**
	 * Returns the width of the rectangle.
	 */
	width() : number {
		return m.rect_width (this.addr);
	}

	/**
	 * Returns the height of the rectangle.	 
	 */
	height() : number {
		return m.rect_height (this.addr);
	}

	/**
	 * Returns `true` if the rectangle is a right rectangle, that is: x1 < x2 and y1 < y2.
	 */
	isRight() : boolean {
		return m.rect_isRight (this.addr);
	}

	/**
	 * Returns the area of the rectangle.
	 * @param strict - Indicates if the area is returned only if the rectangle is a right rectangle.
	 */
	area (strict?: boolean) : number {
		return m.rect_area (this.addr, strict);
	}

	/**
	 * Performs `floor` on the coordinates of the rectangle.
	 */
	floor() : Rect {
		m.rect_floor (this.addr);
		return this;
	}

	/**
	 * Performs `ceil` on the coordinates of the rectangle.
	 */
	ceil() : Rect {
		m.rect_ceil (this.addr);
		return this;
	}

	/**
	 * Truncates the rectangle coordinates to their integer parts.
	 */
	trunc() : Rect {
		m.rect_trunc (this.addr);
		return this;
	}

	/**
	 * Truncates the rectangle coordinates to their fractional parts.
	 */
	fract() : Rect {
		m.rect_fract (this.addr);
		return this;
	}


	/**
	 * Returns the string representation of the rectangle.
	 */
	toString() : string
	{
		return `(${this.x1()}, ${this.y1()}, ${this.x2()}, ${this.y2()})`;
	}
};

File: packages/froxel-math/src/rect.cpp

#include <cstring>
#include <cstdlib>
#include <cmath>
#include <limits>
#include <wasm>

#include "rect.h"
#include "vec2.h"


const double DBL_NULL = std::numeric_limits<double>::max();

export rect *rect_alloc4f (double x1, double y1, double x2, double y2) {
	rect *r = (rect*)malloc(sizeof(rect));
	rect_set4f(r, x1, y1, x2, y2);
	return r;
}

export rect *rect_alloc2f (double width, double height, bool topleft) {
	rect *r = rect_alloc4f(0.0, 0.0, 0.0, 0.0);
	rect_resize(r, width, height, topleft, false);
	return r;
}

export rect *rect_materialize (void *addr) {
	return (rect*)addr;
}

export void rect_free (rect *self) {
	free(self);
}

export rect *rect_clone (rect *self) {
	return rect_alloc4f(self->x1, self->y1, self->x2, self->y2);
}

export void rect_zero (rect *self) {
	memset(self, 0, sizeof(rect));
}

export void rect_reset (rect *self) {
	self->cx = self->cy = DBL_NULL;
	self->x1 = self->y1 = DBL_NULL;
	self->x2 = self->y2 = DBL_NULL;
}

export void rect_extend2f (rect *self, double x, double y)
{
	if (self->x1 == DBL_NULL || x < self->x1) self->x1 = x;
	if (self->y1 == DBL_NULL || y < self->y1) self->y1 = y;
	if (self->x2 == DBL_NULL || x > self->x2) self->x2 = x;
	if (self->y2 == DBL_NULL || y > self->y2) self->y2 = y;

	self->cx = (self->x1 + self->x2) / 2;
	self->cy = (self->y1 + self->y2) / 2;
}

export void rect_extend1v (rect *self, vec2 *v) {
	rect_extend2f(self, v->x, v->y);
}

export void rect_translate2f (rect *self, double dx, double dy)
{
	self->x1 += dx; self->y1 += dy;
	self->x2 += dx; self->y2 += dy;
	self->cx += dx; self->cy += dy;
}

export void rect_translate1v (rect *self, vec2 *v) {
	rect_translate2f(self, v->x, v->y);
}

export void rect_center (rect *self, double x, double y, bool normalized)
{
	if (normalized) {
		x = self->x1 + x*(self->x2 - self->x1);
		y = self->y1 + y*(self->y2 - self->y1);
	}

	rect_translate2f (self, x - self->cx, y - self->cy);
}

export void rect_set4f (rect *self, double x1, double y1, double x2, double y2)
{
	self->cx = (x1 + x2) / 2;
	self->cy = (y1 + y2) / 2;

	self->x1 = x1;
	self->y1 = y1;
	self->x2 = x2;
	self->y2 = y2;
}

export void rect_set1r (rect *self, rect *r) {
	rect_set4f(self, r->x1, r->y1, r->x2, r->y2);
}

export bool rect_equals4f (rect *self, double x1, double y1, double x2, double y2) {
	return x1 == self->x1 && x2 == self->x2 && y1 == self->y1 && y2 == self->y2;
}

export bool rect_equals1r (rect *self, rect *r) {
	return rect_equals4f(self, r->x1, r->y1, r->x2, r->y2);
}

export bool rect_contains4f (rect *self, double x1, double y1, double x2, double y2) {
	return (x1 == std::fmax(self->x1, x1)) && (y1 == std::fmax(self->y1, y1)) && (x2 == std::fmin(self->x2, x2)) && (y2 == std::fmin(self->y2, y2));
}

export bool rect_contains1r (rect *self, rect *r) {
	return rect_contains4f(self, r->x1, r->y1, r->x2, r->y2);
}

export bool rect_contains2f (rect *self, double x, double y, double epsilon) {
	return (self->x1-epsilon <= x && x <= self->x2+epsilon) && (self->y1-epsilon <= y && y <= self->y2+epsilon);
}

export bool rect_contains1v (rect *self, vec2 *v, double epsilon) {
	return rect_contains2f(self, v->x, v->y, epsilon);
}

export void rect_union4f (rect *self, double x1, double y1, double x2, double y2)
{
	self->x1 = std::fmin(self->x1, x1);
	self->y1 = std::fmin(self->y1, y1);
	self->x2 = std::fmax(self->x2, x2);
	self->y2 = std::fmax(self->y2, y2);
}

export void rect_union1r (rect *self, rect *r) {
	rect_union4f(self, r->x1, r->y1, r->x2, r->y2);
}

export bool rect_intersects4f (rect *self, double x1, double y1, double x2, double y2)
{
	x1 = std::fmax(self->x1, x1);
	y1 = std::fmax(self->y1, y1);
	x2 = std::fmin(self->x2, x2);
	y2 = std::fmin(self->y2, y2);

	return std::fmax(0, y2-y1) * std::fmax(0, x2-x1) > 0;
}

export bool rect_intersects1r (rect *self, rect *r) {
	return rect_intersects4f(self, r->x1, r->y1, r->x2, r->y2);
}

export bool rect_intersection4f (rect *self, double x1, double y1, double x2, double y2)
{
	self->x1 = std::fmax(self->x1, x1);
	self->y1 = std::fmax(self->y1, y1);
	self->x2 = std::fmin(self->x2, x2);
	self->y2 = std::fmin(self->y2, y2);

	return std::fmax(0, self->y2-self->y1) * std::fmax(0, self->x2-self->x1) > 0;
}

export bool rect_intersection1r (rect *self, rect *r) {
	return rect_intersection4f(self, r->x1, r->y1, r->x2, r->y2);
}

export void rect_resize (rect *self, double width, double height, bool topLeft, bool normalized)
{
	if (normalized) {
		width *= (self->x2 - self->x1);
		height *= (self->y2 - self->y1);
	}

	if (topLeft) {
		self->x2 = self->x1 + width;
		self->y2 = self->y1 + height;
	}
	else {
		width /= 2;
		height /= 2;

		self->x1 = self->cx - width; self->y1 = self->cy - height;
		self->x2 = self->cx + width; self->y2 = self->cy + height;
	}
}

export void rect_resizeBy (rect *self, double dWidth, double dHeight, bool topLeft)
{
	if (topLeft) {
		self->x2 += dWidth;
		self->y2 += dHeight;
	}
	else {
		dWidth /= 2; dHeight /= 2;

		self->x1 -= dWidth; self->y1 -= dHeight;
		self->x2 += dWidth; self->y2 += dHeight;
	}
}

export double rect_width (rect *self) {
	return self->x2 - self->x1;
}

export double rect_height (rect *self) {
	return self->y2 - self->y1;
}

export bool rect_isRight (rect *self) {
	return self->x1 <= self->x2 && self->y1 <= self->y2;
}

export double rect_area (rect *self, bool strict) {
	return strict ? (rect_isRight(self) ? (self->y2-self->y1)*(self->x2-self->x1) : 0) : ((self->y2-self->y1)*(self->x2-self->x1));
}

export void rect_floor (rect *self)
{
	self->x1 = std::floor(self->x1);
	self->y1 = std::floor(self->y1);
	self->x2 = std::floor(self->x2);
	self->y2 = std::floor(self->y2);
	self->cx = std::floor(self->cx);
	self->cy = std::floor(self->cy);
}

export void rect_ceil (rect *self)
{
	self->x1 = std::ceil(self->x1);
	self->y1 = std::ceil(self->y1);
	self->x2 = std::ceil(self->x2);
	self->y2 = std::ceil(self->y2);
	self->cx = std::ceil(self->cx);
	self->cy = std::ceil(self->cy);
}

export void rect_trunc (rect *self)
{
	self->x1 = (int)(self->x1);
	self->y1 = (int)(self->y1);
	self->x2 = (int)(self->x2);
	self->y2 = (int)(self->y2);
	self->cx = (int)(self->cx);
	self->cy = (int)(self->cy);
}

export void rect_fract (rect *self)
{
	self->x1 -= (int)(self->x1);
	self->y1 -= (int)(self->y1);
	self->x2 -= (int)(self->x2);
	self->y2 -= (int)(self->y2);
	self->cx -= (int)(self->cx);
	self->cy -= (int)(self->cy);
}
